File: Reverse Nodes in k-Group LC-25.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
  
    def reverse(self, head: ListNode, k: int):
      
        prev=None
        curr=next=head
        
        list_tail = head
      
        while k:
          next=curr.next
          curr.next=prev
          prev=curr
          curr=next

          k = k - 1

        list_tail.next = curr 
        list_head = prev
        return list_head, list_tail
  
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
      
        main_head=None
        
        num=0
        slow=fast=head
        
        while fast and fast.next:
          slow=slow.next
          fast=fast.next.next
          num+=1
        
        if not fast:
          num = num * 2
        elif not fast.next:
          num = num * 2 + 1
          
        num=num-(num%k)-k
        
        main_head, main_tail = self.reverse(head, k)
        
        while num>0:
          num = num - k
          main_tail.next, main_tail=self.reverse(main_tail.next,k)
        
        return main_head
          
          
          
          
          
          
          
          
          
          
          
          
          
          
          
          
          
          
          
          
          
          

File: Reverse Nodes in k-Group LC-25.cpp
// My Way
// tc=o(n)
// sc=o(1)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        int len=0;
        ListNode *temp=head;
        while(temp){
            temp=temp->next;
            len++;
        }
        temp=head;
        ListNode* temp1=NULL;
        ListNode *prev=NULL, *curr=head, *next=head;
        for(int i=len;i>=k;i-=k){
            for(int j=0;j<k;j++){
                next=next->next;
                curr->next=prev;
                prev=curr;
                curr=next;
            }
            if(len==i) head=prev;
            if(temp1)temp1->next=prev;
            temp1=temp;
            temp=curr;
            prev=NULL;
        }
        temp1->next=curr;
        return head;
    }
};

// Striver way
// just swap each lines by storing previous
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        int len=0;
        ListNode *dummy=new ListNode();
        dummy->next=head;
        while(head){
            head=head->next;
            len++;
        }
        head=dummy->next;
        ListNode *prev=dummy, *curr, *next;
        while(len>=k){
            curr=prev->next;
            next=curr->next;
            for(int j=1;j<k;j++){
                curr->next=next->next;
                next->next=prev->next;
                prev->next=next;
                next=curr->next;
            }
            len-=k;
            prev=curr;
        }
        return dummy->next;
    }
};